src/hp/magicsock/derp_actor.rs

```diff
@@ -0,0 +1,810 @@
+    time::{Duration, Instant},
+};
+
+use backoff::backoff::Backoff;
+use bytes::{Bytes, BytesMut};
+use tokio::{sync::mpsc, time};
+use tokio_util::sync::CancellationToken;
+use tracing::{debug, info, trace, warn};
+
+use crate::{
+    hp::{
+        derp::{self, DerpMap, MAX_PACKET_SIZE},
+        key::{self, node::PUBLIC_KEY_LENGTH},
+    },
+    inc,
+    metrics::magicsock::MagicsockMetrics,
+    record,
+};
+
+use super::conn::{ActorMessage, Inner};
+
+/// How long a non-home DERP connection needs to be idle (last written to) before we close it.
+const DERP_INACTIVE_CLEANUP_TIME: Duration = Duration::from_secs(60);
+
+/// How often `clean_stale_derp` runs when there are potentially-stale DERP connections to close.
+const DERP_CLEAN_STALE_INTERVAL: Duration = Duration::from_secs(15);
+
+pub(super) enum DerpActorMessage {
+    Send {
+        region_id: u16,
+        contents: Vec<Bytes>,
+        peer: key::node::PublicKey,
+    },
+    Connect {
+        region_id: u16,
+        peer: Option<key::node::PublicKey>,
+    },
+    CloseOrReconnect {
+        region_id: u16,
+        reason: &'static str,
+    },
+    RemovePeerRoutes {
+        peers: Vec<key::node::PublicKey>,
+        region: u16,
+        derp_client: derp::http::Client,
+    },
+    AddPeerRoutes {
+        peers: Vec<key::node::PublicKey>,
+        region: u16,
+        derp_client: derp::http::Client,
+    },
+    NotePreferred(u16),
+    SetDerpMap(Option<DerpMap>),
+    MaybeCloseDerpsOnRebind(Vec<IpAddr>),
+    Shutdown,
+}
+
+/// Contains fields for an active DERP connection.
+#[derive(Debug)]
+struct ActiveDerp {
+    c: derp::http::Client,
+    cancel: CancellationToken,
+    /// The time of the last request for its write
+    /// channel (currently even if there was no write).
+    last_write: Instant,
+    create_time: Instant,
+}
+
+/// A route entry for a public key, saying that a certain peer should be available at DERP
+/// node derpID, as long as the current connection for that derpID is dc. (but dc should not be
+/// used to write directly; it's owned by the read/write loops)
+#[derive(Debug)]
+struct DerpRoute {
+    derp_id: u16,
+    dc: derp::http::Client, // don't use directly; see comment above
+}
+
+pub(super) struct DerpActor {
+    conn: Arc<Inner>,
+    receiver: mpsc::Receiver<DerpActorMessage>,
+    /// DERP regionID -> connection to a node in that region
+    active_derp: HashMap<u16, ActiveDerp>,
+    /// Contains optional alternate routes to use as an optimization instead of
+    /// contacting a peer via their home DERP connection.  If they sent us a message
+    /// on a different DERP connection (which should really only be on our DERP
+    /// home connection, or what was once our home), then we remember that route here to optimistically
+    /// use instead of creating a new DERP connection back to their home.
+    derp_route: HashMap<key::node::PublicKey, DerpRoute>,
+    msg_sender: flume::Sender<ActorMessage>,
```

Comment:
This is to send messages to some actor. But not to self, because that would be `DerpActorMessage`. Maybe it would be helpful to prefix the namespace.

Also, why are some actor connections using flume and some tokio mpsc?